app/src/rgb_underglow.c

```diff
@@ -313,63 +335,70 @@ int zmk_rgb_underglow_cycle_effect(int direction) {
 }
 
 int zmk_rgb_underglow_toggle() {
-    if (!led_strip)
-        return -ENODEV;
+    return state.on ? zmk_rgb_underglow_off() : zmk_rgb_underglow_on();
+}
+
+int zmk_rgb_underglow_set_hsb(struct zmk_led_hsb color) {
+    if (color.h > 360 || color.s > 100 || color.b > 100) {
+        return -ENOTSUP;
+    }
 
-    state.on = !state.on;
+    state.color = color;
 
-#if IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_EXT_POWER)
-    if (ext_power != NULL) {
-        int rc;
+    return 0;
+}
 
-        if (state.on) {
-            rc = ext_power_enable(ext_power);
-        } else {
-            rc = ext_power_disable(ext_power);
-        }
+struct zmk_led_hsb zmk_rgb_underglow_calc_hue(int direction) {
+    struct zmk_led_hsb color = state.color;
 
-        if (rc != 0) {
-            LOG_ERR("Unable to toggle EXT_POWER: %d", rc);
-        }
+    if (color.h == 0 && direction < 0) {
+        return color;
     }
-#endif
 
-    if (state.on) {
-        state.animation_step = 0;
-        k_timer_start(&underglow_tick, K_NO_WAIT, K_MSEC(50));
-    } else {
-        zmk_rgb_underglow_off();
+    color.h += direction * CONFIG_ZMK_RGB_UNDERGLOW_HUE_STEP;
+
+    color.h = color.h % 360;
 
-        k_timer_stop(&underglow_tick);
+    return color;
+}
+
+struct zmk_led_hsb zmk_rgb_underglow_calc_sat(int direction) {
+    struct zmk_led_hsb color = state.color;
+
+    if (color.s == 0 && direction < 0) {
```

Comment:
```suggestion
    if (color.s - CONFIG_ZMK_RGB_UNDERGLOW_SAT_STEP < 0 && direction < 0) {
```
Something like this is better. Or even setting to zero if it's less than 0 after the fact. If the step is more than 1, this won't catch all cases (which I doubt many people would use 1).